File: mall-all/src/main/java/com/cyh/mallportal/entity/Order.java
package com.cyh.mallportal.entity;

import com.baomidou.mybatisplus.annotation.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * 订单实体类
 * 对应数据库表 orders
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@TableName(value = "orders", autoResultMap = true)
public class Order implements Serializable {

    /**
     * 主键ID
     */
    @TableId(value = "id", type = IdType.AUTO)
    private Long id;

    /**
     * 订单号
     */
    @TableField(value = "order_no")
    private String orderNo;

    /**
     * 用户ID
     */
    @TableField(value = "user_id")
    private Long userId;

    /**
     * 订单总金额
     */
    @TableField(value = "total_amount")
    private BigDecimal totalAmount;

    /**
     * 实付金额
     */
    @TableField(value = "pay_amount")
    private BigDecimal payAmount;

    /**
     * 优惠金额
     */
    @TableField(value = "discount_amount")
    private BigDecimal discountAmount;

    /**
     * 运费
     */
    @TableField(value = "freight_amount")
    private BigDecimal freightAmount;

    /**
     * 订单状态: 1-待付款 2-待发货 3-待收货 4-已完成 5-已取消 6-退款中 7-已退款 8-已拒绝
     */
    @TableField(value = "status")
    private Integer status;

    /**
     * 支付状态: 0-未支付 1-已支付 2-已退款
     */
    @TableField(value = "pay_status")
    private Integer payStatus;

    /**
     * 支付时间
     */
    @TableField(value = "pay_time")
    private LocalDateTime payTime;

    /**
     * 支付截止时间（下单后超时未支付自动取消）
     */
    @TableField(value = "expire_time")
    private LocalDateTime expireTime;

    /**
     * 支付方式: alipay-支付宝 wechat-微信
     */
    @TableField(value = "pay_type")
    private String payType;

    /**
     * 配送方式
     */
    @TableField(value = "delivery_type")
    private String deliveryType;

    /**
     * 物流公司
     */
    @TableField(value = "delivery_company")
    private String deliveryCompany;

    /**
     * 物流单号
     */
    @TableField(value = "delivery_no")
    private String deliveryNo;

    /**
     * 发货时间
     */
    @TableField(value = "delivery_time")
    private LocalDateTime deliveryTime;

    /**
     * 收货时间
     */
    @TableField(value = "receive_time")
    private LocalDateTime receiveTime;

    /**
     * 收货人
     */
    @TableField(value = "receiver_name")
    private String receiverName;

    /**
     * 收货人电话
     */
    @TableField(value = "receiver_phone")
    private String receiverPhone;

    /**
     * 收货地址
     */
    @TableField(value = "receiver_address")
    private String receiverAddress;

    /**
     * 备注
     */
    @TableField(value = "remark")
    private String remark;

    /**
     * 创建时间
     */
    @TableField(value = "created_at", fill = FieldFill.INSERT)
    private LocalDateTime createdAt;

    /**
     * 更新时间
     */
    @TableField(value = "updated_at", fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updatedAt;

    /**
     * 取消原因
     */
    @TableField(value = "cancel_reason")
    private String cancelReason;

    /**
     * 乐观锁版本号
     */
    @Version
    @TableField(value = "version")
    private Integer version;

    /**
     * 逻辑删除: 0-正常 1-已删除（用户删除标记，非MP逻辑删除）
     * 用户查询手动过滤 is_deleted=0，商家/管理员不对此字段做过滤
     */
    private Integer isDeleted;

    /**
     * 退款原因
     */
    @TableField(value = "refund_reason")
    private String refundReason;

    /**
     * 退款金额
     */
    @TableField(value = "refund_amount")
    private BigDecimal refundAmount;

    /**
     * 退款来源状态: 2-待发货 3-待收货 4-已完成
     * 申请退款时记录退款前的状态，取消退款时恢复
     */
    @TableField(value = "refund_from_status")
    private Integer refundFromStatus;

    /**
     * 退款操作人ID（审核通过或拒接退款的操作用户ID）
     */
    @TableField(value = "refund_operator_id")
    private Long refundOperatorId;

    /**
     * 拒绝原因（拒接退款时填写）
     */
    @TableField(value = "reject_reason")
    private String rejectReason;

    /**
     * 拒绝时间（拒接退款时记录）
     */
    @TableField(value = "rejected_at")
    private LocalDateTime rejectedAt;
}

File: mall-frontend/ecommerce-frontend/src/api/order.ts
import request from '@/utils/request'

/**
 * 订单项接口（订单详情用）
 */
export interface OrderItem {
  productName: string
  productImage: string
  skuSpecs: string
  price: number
  quantity: number
  totalAmount: number
}

/**
 * 发货记录接口
 */
export interface Delivery {
  deliveryCompany: string
  deliveryNo: string
  deliveryStatus: number
  deliveryTime: string
}

/**
 * 订单详情响应（扁平结构，新 API 格式）
 */
export interface OrderDetail {
  id: number
  orderNo: string
  totalAmount: number
  payAmount: number
  discountAmount: number
  freightAmount: number
  status: number
  payStatus: number
  statusDesc: string
  payStatusDesc: string
  payTypeDesc: string
  payTime: string | null
  expireTime: string | null
  receiveTime: string | null
  receiverName: string
  receiverPhone: string
  receiverAddress: string
  remark: string
  createdAt: string
  cancelReason: string | null
  refundReason: string | null
  refundAmount: number | null
  rejectReason: string | null
  rejectedAt: string | null
  items: OrderItem[]
  deliveries: Delivery[]
}

/**
 * 订单接口（订单列表用）
 */
export interface Order {
  id: number
  orderNo: string
  userId: number
  totalAmount: number
  payAmount: number
  discountAmount: number
  freightAmount: number
  status: number
  payStatus: number
  payTime: string | null
  payType: string | null
  deliveryCompany: string | null
  deliveryNo: string | null
  deliveryTime: string | null
  receiveTime: string | null
  receiverName: string
  receiverPhone: string
  receiverAddress: string
  expireTime: string | null
  refundAmount: number | null
  remark: string
  createdAt: string
  updatedAt: string
  cancelReason: string | null
  rejectReason?: string
  rejectedAt?: string
  version: number
  statusDesc?: string
  itemCount?: number
  totalQuantity?: number
  items?: OrderItem[]
  shippingInfo?: {
    company: string
    trackingNo: string
  }
}

/**
 * 分页响应
 */
export interface PageResult<T> {
  list: T[]
  page: number
  pageSize: number
  total: number
}

/**
 * 创建订单请求
 */
export interface CreateOrderRequest {
  addressId?: number
  receiverName?: string
  receiverPhone?: string
  receiverAddress?: string
  totalAmount: number
  discountAmount?: number
  freightAmount?: number
  payAmount: number
  remark?: string
  items: Array<{
    skuId: number
    spuId: number
    quantity: number
    productName: string
    productImage?: string
    skuSpecs?: string
    price: number
    giftFlag?: number
  }>
}

/**
 * 从购物车创建订单请求
 */
export interface CreateOrderFromCartRequest {
  addressId: number
  payType: string
  buyerMessage?: string
}

/**
 * 创建订单响应
 */
export interface CreateOrderResponse {
  orderId: number
  orderNo: string
  payAmount: number
  totalAmount?: number
  order?: {
    id: number
    orderNo: string
    totalAmount?: number
    payAmount?: number
  }
}

// 创建订单
export function createOrder(data: CreateOrderRequest): Promise<CreateOrderResponse> {
  return request({
    url: '/order/create',
    method: 'post',
    data
  })
}

// 从购物车结算创建订单（后端使用 @RequestParam，返回 List<Long> 即 [orderId]）
export function createOrderFromCart(data: CreateOrderFromCartRequest): Promise<number[]> {
  const params: Record<string, any> = { addressId: data.addressId }
  if (data.buyerMessage) params.buyerMessage = data.buyerMessage
  return request({
    url: '/order/create-from-cart',
    method: 'post',
    params
  })
}

// 获取订单详情
export function getOrderDetail(orderNo: string): Promise<OrderDetail> {
  return request({
    url: `/order/detail/${orderNo}`,
    method: 'get'
  })
}

// 获取订单列表（分页，带商品明细）
export function getOrderListWithItems(params?: {
  page?: number
  pageSize?: number
  status?: number
}): Promise<PageResult<Order>> {
  const cleanParams: Record<string, any> = {}
  if (params) {
    Object.entries(params).forEach(([key, val]) => {
      if (val !== undefined && val !== null) cleanParams[key] = val
    })
  }
  return request({
    url: '/order/list-with-items',
    method: 'get',
    params: cleanParams
  })
}

// 获取订单列表（分页）
export function getOrderList(params?: { 
  status?: number, 
  page?: number, 
  pageSize?: number 
}): Promise<PageResult<Order>> {
  // 清理 undefined 参数，防止 Axios 将其序列化为 "undefined" 字符串
  const cleanParams: Record<string, any> = {}
  if (params) {
    Object.entries(params).forEach(([key, val]) => {
      if (val !== undefined && val !== null) {
        cleanParams[key] = val
      }
    })
  }
  return request({
    url: '/order/list',
    method: 'get',
    params: cleanParams
  })
}

// 获取订单明细
export function getOrderItems(orderId: number): Promise<OrderItem[]> {
  return request({
    url: `/order/items/${orderId}`,
    method: 'get'
  })
}

// 管理员分页查询全部订单列表
export function getAdminOrderList(params?: {
  status?: number
  userId?: number
  orderNo?: string
  payTimeStart?: string
  payTimeEnd?: string
  deliveryTimeStart?: string
  deliveryTimeEnd?: string
  receiveTimeStart?: string
  receiveTimeEnd?: string
  page?: number
  pageSize?: number
}): Promise<PageResult<Order>> {
  // 清理 undefined 参数
  const cleanParams: Record<string, any> = {}
  if (params) {
    Object.entries(params).forEach(([key, val]) => {
      if (val !== undefined && val !== null) {
        cleanParams[key] = val
      }
    })
  }
  return request({
    url: '/order/admin/list',
    method: 'get',
    params: cleanParams
  })
}

// 商家查询店铺订单列表
export function getSellerOrderList(params?: {
  status?: number
  page?: number
  pageSize?: number
  userId?: number
  orderNo?: string
  payTimeStart?: string
  payTimeEnd?: string
  deliveryTimeStart?: string
  deliveryTimeEnd?: string
  receiveTimeStart?: string
  receiveTimeEnd?: string
}): Promise<{ list: any[]; total: number; page: number; pageSize: number }> {
  const cleanParams: Record<string, any> = {}
  if (params) {
    Object.entries(params).forEach(([key, val]) => {
      if (val !== undefined && val !== null && val !== '') {
        cleanParams[key] = val
      }
    })
  }
  return request({
    url: '/order/seller/list',
    method: 'get',
    params: cleanParams
  })
}

// 商家查询订单详情
export function getSellerOrderDetail(orderNo: string): Promise<OrderDetail> {
  return request({
    url: `/order/seller/detail/${orderNo}`,
    method: 'get'
  })
}

// 取消订单
export function cancelOrder(orderId: number): Promise<void> {
  return request({
    url: `/order/cancel/${orderId}`,
    method: 'put'
  })
}

// 支付订单
export function payOrder(orderId: number, payType?: string): Promise<void> {
  return request({
    url: `/order/pay/${orderId}`,
    method: 'put',
    params: payType ? { payType } : undefined
  })
}

// 批量付款（从购物车结算后调用）
export interface BatchPayRequest {
  orderIds: number[]
  payType?: string
}

export interface BatchPayFailItem {
  orderId: number
  reason: string
}

export interface BatchPayResponse {
  success: number[]
  fail: BatchPayFailItem[]
  totalCount: number
  successCount: number
  failCount: number
}

export function batchPay(data: BatchPayRequest): Promise<BatchPayResponse> {
  return request({
    url: '/order/batch-pay',
    method: 'post',
    data
  })
}

// 订单发货
export function deliverOrder(orderId: number, data: { 
  deliveryCompany: string, 
  deliveryNo: string 
}): Promise<void> {
  return request({
    url: `/order/deliver/${orderId}`,
    method: 'put',
    params: data
  })
}

// 确认收货
export function confirmReceive(orderId: number): Promise<void> {
  return request({
    url: `/order/confirm/${orderId}`,
    method: 'put'
  })
}

// 删除订单
export function deleteOrder(orderId: number): Promise<void> {
  return request({
    url: `/order/delete/${orderId}`,
    method: 'delete'
  })
}

// 取消退款申请
export function cancelRefund(orderId: number): Promise<void> {
  return request({
    url: `/order/refund/cancel/${orderId}`,
    method: 'put'
  })
}

// 用户申请退款
export function applyRefund(orderId: number, refundReason?: string): Promise<void> {
  return request({
    url: `/order/refund/apply/${orderId}`,
    method: 'post',
    params: refundReason ? { refundReason } : undefined
  })
}

// 获取订单状态描述
export function getStatusDesc(status: number): Promise<{ statusDesc: string }> {
  return request({
    url: `/order/status-desc/${status}`,
    method: 'get'
  })
}

// 获取支付状态描述
export function getPayStatusDesc(payStatus: number): Promise<{ payStatusDesc: string }> {
  return request({
    url: `/order/pay-status-desc/${payStatus}`,
    method: 'get'
  })
}

// 管理员审核退款
export function refundReview(orderId: number, reviewResult: number, rejectReason?: string): Promise<void> {
  return request({
    url: `/order/refund/review/${orderId}`,
    method: 'put',
    params: { reviewResult, ...(rejectReason ? { rejectReason } : {}) }
  })
}

// 同意退款
export function refundApprove(orderId: number): Promise<void> {
  return request({
    url: `/order/refund/approve/${orderId}`,
    method: 'put'
  })
}

// 拒绝退款
export function refundReject(orderId: number, rejectReason: string): Promise<void> {
  return request({
    url: `/order/refund/reject/${orderId}`,
    method: 'put',
    params: { rejectReason }
  })
}

// 订单状态统计接口
export interface OrderStatusCount {
  pendingPayment: number    // 待付款
  pendingDelivery: number   // 待发货
  pendingReceipt: number    // 待收货
  refunding: number         // 退款中
}

// 获取订单状态统计（用户个人中心）
export function getOrderStatusCount(): Promise<OrderStatusCount> {
  return request({
    url: '/order/count-status',
    method: 'get'
  })
}